Creds-init writes to fixed location when HOME override is disabled

sbwsg opened the pull request:
# Changes

Fixes #2165
Relates to https://github.com/tektoncd/pipeline/issues/2013

When the disable-home-env-overwrite flag is set to "true" each Step in a Task can conceivably have its own HOME directory. The concept of "HOME" is further muddied in systems that randomize the UID of containers.

When the disanle-home-env-overwite flag is set to "true", creds-init will write its credentials to `/tekton/creds` instead of `/tekton/home`.  Then the entrypoint binary will copy the credentials from `/tekton/creds` into `$HOME` before the Task's Steps are allowed to run.  This change should be completely transparent to the user.

For those users who were for some very specific reason actually relying on the credentials being in `/tekton/home` we'll now expose a new variable substitution: `$(credentials.path)` will get replaced with the directory that creds-init writes to - either `/tekton/home` or `/tekton/creds` depending on the state of the disable-home-env-overwrite flag.

This is a stop-gap measure as we work towards beta. It seems that we will probably need to revisit our recommendations around auth and the contract we're making with creds-init.

- [x] Variable replacement for `$(credentials.path)`

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) to build and release this image.

## Reviewer Notes

If [API changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md) are included, [additive changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#additive-changes) must be approved by at least two [OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS) and [backwards incompatible changes](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-incompatible-changes) must be approved by [more than 50% of the OWNERS](https://github.com/tektoncd/pipeline/blob/master/OWNERS), and they must first be added [in a backwards compatible way](https://github.com/tektoncd/pipeline/blob/master/api_compatibility_policy.md#backwards-compatible-changes-first).

# Release Notes

```
Users who disable Tekton's HOME env overwrite will now find that git and docker credentials are written to a volumeMount, /tekton/creds, before they're copied into $HOME by the entrypoint. A new variable replacement has been introduced, $(credentials.path), which expands to the directory where creds-init writes its credentials, either /tekton/home or /tekton/creds.
```

sbwsg commented:
/hold

Want to look at this again with fresh eyes on Monday before committing.

vdemeester commented on docs/auth.md:
```diff
@@ -24,6 +24,12 @@ To solve this, before any `PipelineResources` are retrieved, all `pods` execute
 a credential initialization process that accesses each of its secrets and
 aggregates them into their respective files in `$HOME`.
 
+Please note: If the `disable-home-env-overwrite` feature flag has been set to
```
So, in the case of `disable-home-env-overwrite` (and by default in the future), do we decide that `/tekton/home` is where we write anything that "would be in `$HOME`" ? (like `.gitconfig`, …) I am +:100: on this as we will document it, and we are sure it's always written at the same place.

sbwsg replied:
Interesting idea!  I think this might make sense.  I was considering using a more explicit name like `/tekton/creds` since we'd originally planned for /tekton/home to be removed in future.  But if we see this as reusable then I think it makes sense to leave it as-is.

I'm going to write up a one-page design doc and will present it to the WG on wednesday just to quickly get community feedback on that plan.

vdemeester replied:
+:100: on both your comments :wink: I really like `/tekton/creds` and having a variable for it is :heart_eyes_cat:

sbwsg replied:
👍 awesome, i've implemented both of these now.

sbwsg commented:
I've written up a very short design doc covering the solution in this PR and a slight alternative.  Doc here: https://docs.google.com/document/d/1SVuDt-SXPHymz41dveSXFSPrx5Z-Wzb9eHliJAyYg4o

sbwsg commented:
Note to self: This would probably break PipelineResources when the `disable-home-env-overwrite` flag is enabled (and eventually flipped to on by default in the following release).  Might need to update PipelineResources to account for this.

sbwsg commented:
Updated to write to /tekton/creds instead of /tekton/home.

Still looking at:
- Variable replacement for `$(credentials.path)`
- Fixing `PipelineResources` to work with the credentials path.
- Some kind of integration testing (though at the moment I might have to postpone this until after the flag gets flipped to be default true... not sure)

sbwsg commented:
I've added variable replacement from `$(credentials.path)` to `/tekton/home` or `/tekton/creds` depending on what the disable-home-env-overwrite flag is set to.

I've updated the `git` PipelineResource to deal with the possibility that creds are in /tekton/creds instead of /tekton/home.

sbwsg commented:
This change should be pretty much invisible to users now, no extra copying in their Steps should be necessary.  The entrypoint binary now copies the creds into $HOME instead of creds-init (when the disable-home-env-overwrite feature flag is true) so the net result is the same regardless of the feature flag: creds end up in container HOME dirs.

sbwsg commented:
/hold cancel

I think this is probably ready for review now.

vdemeester commented on cmd/entrypoint/main.go:
```diff
@@ -53,6 +54,13 @@ func main() {
+
+	// Copy any creds injected by creds-init into the $HOME directory of the current
+	// user so that they're discoverable by git / ssh.
+	if err := credentials.CopyCredsToHome(credentials.CredsInitCredentials); err != nil {
```
any reason why it is required here ? (in the `entrypoint` wrapper binary)

sbwsg replied:
`entrypoint` is the only tool we have that for-sure runs in the Step container, with access to (and knowledge of) the Step's HOME dir.

So we can only really use entrypoint to perform this copy I think.  And the nice benefit is that the behaviour now matches precisely before and after the feature flag: before the flag the credentials end up in $HOME (creds-init writes them there) and after the flag the credentials end up in $HOME (entrypoint writes them there).

vdemeester replied:
ah very good point :+1: gotcha :wink:

vdemeester commented on pkg/reconciler/taskrun/resources/apply.go:
```diff
@@ -125,6 +125,15 @@ func ApplyTaskResults(spec *v1alpha1.TaskSpec) *v1alpha1.TaskSpec {
 
+// ApplyCredentialsPath applies a substitution of the key $(credentials.path) with the path that the creds-init
+// helper will write its credentials to.
+func ApplyCredentialsPath(spec *v1alpha1.TaskSpec, path string) *v1alpha1.TaskSpec {
```
Is this only usable as internal ? (because I think it's gonna be invalid if using directly in a spec)

sbwsg replied:
Could you tell me more about it being invalid if used directly in spec?  It's intended for users to be able to work with as well.

vdemeester replied:
I think this might fail at validation saying `credential.path` is an invalid variable (not 100% sure but my feeling is that not having anything about it in the validation code means it will fail — maybe I am wrong though, a tests showing this would be nice :stuck_out_tongue_closed_eyes: )

sbwsg replied:
Aaaaah.  Lemme check.  Great catch.

sbwsg replied:
Initial manual test proved positive - I was able to use `$(credential.path)` in a TaskRun w/ embedded TaskSpec.  I didn't hit a validation error and the Task ran successfully.

I'm going to add a test for this right now to make sure it doesn't fail validation in future.

sbwsg replied:
Updated with tests for task and taskrun validations in both v1beta1 and v1alpha1 packages.

vdemeester commented:
/test pull-tekton-pipeline-integration-tests

vdemeester commented:
/test pull-tekton-pipeline-integration-tests

vdemeester commented:
/test pull-tekton-pipeline-build-tests
